cmd/agent/app/session.go

```diff
@@ -1,230 +1,160 @@
-	// 	}
-	// }
-
-	// return activeConnections
-
-	return session.connections
-}
-
-func NewSession(id string, juicePath string, version string, gpus *gpu.SelectedGpuSet, eventListener EventListener, persistent bool) *Session {
-	session := Session{
-		id:            id,
+func newSession(ctx context.Context, id string, version string, persistent bool, juicePath string, gpus *gpu.SelectedGpuSet, eventListener EventListener) *Session {
+	return &Session{
+		Id:            id,
+		Version:       version,
+		Persistent:    persistent,
 		juicePath:     juicePath,
-		version:       version,
-		state:         restapi.SessionActive,
 		gpus:          gpus,
-		persistant:    persistent,
-		connections:   orderedmap.New[string, *Reference[connection.Connection]](),
+		connections:   utilities.NewConcurrentMap[string, *Connection](),
+		taskManager:   task.NewTaskManager(ctx),
 		eventListener: eventListener,
 	}
-
-	return &session
 }
 
 func (session *Session) Session() restapi.Session {
-	session.mutex.RLock()
-	defer session.mutex.RUnlock()
+	return utilities.WithReturn(session.closed, func(value bool) restapi.Session {
+		connections := make([]restapi.Connection, 0, session.connections.Len())
+		gpus := make([]restapi.SessionGpu, 0)
+		state := restapi.SessionClosed
+
+		if !value {
+			session.connections.Foreach(func(key string, value *Connection) bool {
+				connections = append(connections, restapi.Connection{
+					ConnectionData: value.ConnectionData,
+				})
+				return true
+			})
 
-	connections := make([]restapi.Connection, 0, session.connections.Len())
-	for pair := session.connections.Oldest(); pair != nil; pair = pair.Next() {
-		connections = append(connections, pair.Value.Object.Connection())
-	}
+			gpus = session.gpus.GetGpus()
+			state = restapi.SessionActive
+		}
 
-	return restapi.Session{
-		Id:          session.id,
-		State:       session.state,
-		Version:     session.version,
-		Gpus:        session.gpus.GetGpus(),
-		Connections: connections,
-		Persistent:  session.persistant,
-	}
+		return restapi.Session{
+			Id:          session.Id,
+			State:       state,
+			Version:     session.Version,
+			Gpus:        gpus,
+			Connections: connections,
+			Persistent:  session.Persistent,
+		}
+	})
 }
 
-func (session *Session) Close() error {
-	errs := []error{}
-	// verify that there are no connections
-	if session.ActiveConnections().Len() > 0 {
-		logger.Errorf("closing session has active connections, %s", session.Id())
-		errs = append(errs, errors.New("closing session has active connections"))
-	}
-	session.mutex.Lock()
+func (session *Session) Run(group task.Group) error {
+	group.GoFn(fmt.Sprintf("session %s close", session.Id), func(g task.Group) error {
+		<-session.taskManager.Ctx().Done()
 
-	session.state = restapi.SessionClosed
+		session.closed.Set(true)
 
-	session.connections = nil
+		err := session.taskManager.Wait()
 
-	session.gpus.Release()
-	session.gpus = nil
+		session.gpus.Release()
```

Comment:
Should we do this only when the session is actually closed by the agent thread?

Reply:
This is the appropriate place. The gpus object here is a subset of the agent gpus. This is basically telling the agent the gpus are available again.